### DISCUSSION TITLE
Вычисляемые кодом префильтры.

### DISCUSSION THREAD

**Discussion by Ekhorikov2** [totumonline/totum-ru-issues-and-discussions#209] [Как это сделать?] [answered]
Всем привет.
Подскажите, пожалуйста, по настройке вычисляемых кодом префильтров.
Задача следующая:
Есть основная, простая таблица,  в которой хранятся записи о доставках заказов. Одно из полей этой таблицы, это Город, куда доставляется заказ!
![image](https://user-images.githubusercontent.com/112487577/187440016-f2279dee-1e36-480f-955b-a17953ef52ee.png)
Есть несколько менеджеров, которые работают, каждый со своим городом. В другой простой таблице, настроены кнопки, в виде меню, на каждый из городов. 
![image](https://user-images.githubusercontent.com/112487577/187439802-6165792a-5680-4a5e-9439-112611c96d19.png)
При нажатии на кнопку города, открывается первая таблица, через префильтр поля Город. Таким образом менеджеры видят одну и ту же таблицу, но заказы в ней только по их городу. Вроде бы всё работает, менеджер нажимает кнопку со своим городом и работает со своими заказами, без возможности выбрать другой город из самой таблицы (префильтр по полю Город скрыт в таблице от пользователей), но вот проблема в том, что если нажать на ластик в блоке префильтров, значения сбрасываются у всех префильтров, и открывается вся таблица, со всеми городами.
Вопрос:
Как рассчитать значение префильра кодом, чтоб его невозможно было сбросить или как либо изменить?
В учебном курсе есть об этом информация, но совсем немного. Подскажите как настроить в моём случае?
Заранее спасибо.
PS: Если возник вопрос как пришли к организации таблицы через кнопки, то основной идеей было в дальнейшем ограничить срабатывание кнопок по условию, по пользователям.

**Comment by Alexmeb** [comment]
)) всё не то

**Reply by Ekhorikov2** [reply]
> )) всё не то

???

**Reply by ttm-alexeypolunin** [reply]
При использовании `linkToTable` нельзя заблокировать Редактирование поля Префильтра (если заблокировать не передадутся значения по зашифрованной адресной строке), но можно сделать его видимым только Создателю (что и было сделано в описании :👆.

**Reply by Ekhorikov2** [reply]
> При использовании `linkToTable` нельзя заблокировать Редактирование поля Префильтра (если заблокировать не передадутся значения по зашифрованной адресной строке), но можно сделать его видимым только Создателю (что и было сделано в описании :👆.

А какой выход? Убрать кнопки и оставить таблицу открытой для всех пользователей, чтоб сами вручную фильтрами?

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Если пользователь должен иметь доступ только к определенным городам, то я бы в таблице пользователей или еще где сделал соответствие пользователя и города (или нескольких городов).

Дальше делается дополнительное поле в префильтре тип `Селект`, привязывается например к `id` с двумя дополнительными значениями `"*ALL*"` и `"*NONE*"`.

Оно невидимое и нередактируемое.

В нем делаем Код проверяющий на допустимый город в поле префильтра с городом. Если проверку на допустимость города для пользователя `$#nu` проходит то `"*ALL*"`, если нет то `"*NONE*"`

**Reply by ttm-alexeypolunin** [reply]
P>S Еще в Префильтре Селекте с городом я бы тоже значением по умолчанию поставил `*NONE*` (и включил его как возможное значение). Одно это на самом деле решит вопрос.

Другое дело, что вы хотите потом кнопки распределить по пользователям, а если через форматирование это сделать, то можно будет обойти. А проверка кодом в префильтре — железная.

**Reply by Ekhorikov2** [reply]
> P>S Еще в Префильтре Селекте с городом я бы тоже значением по умолчанию поставил `*NONE*` (и включил его как возможное значение). Одно это на самом деле решит вопрос.
> 
> Другое дело, что вы хотите потом кнопки распределить по пользователям, а если через форматирование это сделать, то можно будет обойти. А проверка кодом в префильтре — железная.

Сейчас скорей разговор за удобство идёт, чтоб менеджеры не тратили время на фильтрацию, чтоб попадание к заявкам было максимально коротким. Ограничения сейчас не главное.
